src/jit/hwintrinsicxarch.cpp

```diff
@@ -353,11 +354,78 @@ GenTree* Compiler::impSSE2Intrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HAN
+        case NI_SSE2_CompareGreaterThan:
+        case NI_SSE2_CompareGreaterThanOrEqual:
+        case NI_SSE2_CompareLessThan:
+        case NI_SSE2_CompareLessThanOrEqual:
+        case NI_SSE2_CompareNotEqual:
+        case NI_SSE2_CompareNotGreaterThan:
+        case NI_SSE2_CompareNotGreaterThanOrEqual:
+        case NI_SSE2_CompareNotLessThan:
+        case NI_SSE2_CompareNotLessThanOrEqual:
+        case NI_SSE2_CompareOrdered:
+        case NI_SSE2_CompareUnordered:
+        case NI_SSE2_ConvertToVector128Int32:
+        case NI_SSE2_ConvertToVector128Double:
+        case NI_SSE2_ConvertToVector128Single:
+        case NI_SSE2_ConvertToVector128Int32WithTruncation:
+        case NI_SSE2_Divide:
+        case NI_SSE2_Max:
+        case NI_SSE2_Min:
+        case NI_SSE2_Multiply:
+        case NI_SSE2_MultiplyHigh:
+        case NI_SSE2_MultiplyLow:
+        case NI_SSE2_MultiplyHorizontalAdd:
+        case NI_SSE2_Or:
+        case NI_SSE2_PackSignedSaturate:
+        case NI_SSE2_PackUnsignedSaturate:
+        case NI_SSE2_SumAbsoluteDifferences:
+        case NI_SSE2_Sqrt:
+        case NI_SSE2_Subtract:
+        case NI_SSE2_SubtractSaturate:
+        case NI_SSE2_UnpackHigh:
+        case NI_SSE2_UnpackLow:
+        case NI_SSE2_Xor:
             assert(sig->numArgs == 2);
             op2      = impSIMDPopStack(TYP_SIMD16);
             op1      = impSIMDPopStack(TYP_SIMD16);
             baseType = getBaseTypeOfSIMDType(sig->retTypeSigClass);
-            retNode  = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, op2, NI_SSE2_Add, baseType, 16);
+            retNode  = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, op2, intrinsic, baseType, 16);
+            break;
+
+
+        case NI_SSE2_ExtractInt16:
+        case NI_SSE2_ExtractUInt16:
+            assert(sig->numArgs == 2);
+            imm8 = impSIMDPopStack(TYP_UBYTE);
+            op1 = impSIMDPopStack(TYP_SIMD16);
+            baseType = getBaseTypeOfSIMDType(sig->retTypeSigClass);
+            retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, imm8, intrinsic, baseType, 16);
+            break;
+
+        case NI_SSE2_InsertInt16:
+            assert(sig->numArgs == 3);
+            imm8 = impSIMDPopStack(TYP_UBYTE);
+            op2 = impSIMDPopStack(TYP_SHORT);
+            op1 = impSIMDPopStack(TYP_SIMD16);
+            baseType = getBaseTypeOfSIMDType(sig->retTypeSigClass);
+            // TODO - need imm8 support
+            retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, op2, intrinsic, baseType, 16);
+            break;
+
+        case NI_SSE2_InsertUInt16:
+            assert(sig->numArgs == 3);
+            imm8 = impSIMDPopStack(TYP_UBYTE);
+            op2 = impSIMDPopStack(TYP_USHORT);
+            op1 = impSIMDPopStack(TYP_SIMD16);
+            baseType = getBaseTypeOfSIMDType(sig->retTypeSigClass);
+            // TODO - need imm8 support
+            retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, op2, intrinsic, baseType, 16);
```

Comment:
@mikedn Do you how to handle immediate values passed to intrinsics (they will compile time constants)
for couple of the following functions
```
public static short ExtractInt16<T>(Vector128<T> value, byte index)
public static Vector128<T> InsertInt16<T>(Vector128<T> value, short data, byte index);
```

Reply:
Considering that only immediate values are needed I'd simply add a integer data member to `GenTreeHWIntrinsic` and store the immediate in it. Changing `GenTreeHWIntrinsic` to actually support 3 operands is overkill and would likely cause problems in certain areas of the JIT that aren't used to deal with more than 2 operands.

FWIW there are 3 operand nodes (e.g. `GenTreeCmpXchg`) but they're rare/special.

Reply:
> Changing GenTreeHWIntrinsic to actually support 3 operands is overkill and would likely cause problems in certain areas of the JIT that aren't used to deal with more than 2 operands.

For intrinsics that have more than three parameters, we can link the arguments in a list (`GT_LIST`).
```
(GTHWIntrinsics op1 (GT_LIST op2 (GT_LIST op3 (GT_LIST op4 null))))
```

Reply:
Yes, using a list is another option but that too is overkill for these immediate operands. Also, in the form that you show it makes things harder for any code that needs to deal with the common case of having 2 operands.